harness/tests/integration_cases/test_raw_node.rs

```diff
@@ -154,50 +156,278 @@ fn test_raw_node_read_index_to_old_leader() {
+            let mut cc = conf_change_v2(vec![]);
+            cc.set_context(context.clone());
+            raw_node.propose_conf_change(vec![], cc).unwrap();
+            rd = raw_node.ready();
+        }
+
+        // Check that the right ConfChange comes out.
+        assert_eq!(rd.entries().len(), 1);
+        assert_eq!(
+            rd.entries()[0].get_entry_type(),
+            EntryType::EntryConfChangeV2
+        );
+        let mut leave_cc = ConfChangeV2::default();
+        leave_cc
+            .merge_from_bytes(rd.entries()[0].get_data())
+            .unwrap();
+        assert_eq!(context, leave_cc.get_context(), "{:?}", cc.as_v2());
+        // Lie and pretend the ConfChange applied. It won't do so because now
+        // we require the joint quorum and we're only running one node.
+        let cs = raw_node.apply_conf_change(&leave_cc).unwrap();
+        assert_eq!(cs, exp2.unwrap());
+    }
+}
+
+/// Tests the configuration change auto leave even leader lost leadership.
+#[test]
+fn test_raw_node_joint_auto_leave() {
+    let l = default_logger();
+
+    let single = new_conf_change_single(2, ConfChangeType::AddLearnerNode);
+    let mut test_cc = conf_change_v2(vec![single]);
+    test_cc.set_transition(ConfChangeTransition::Implicit);
+    let exp_cs = conf_state_v2(vec![1], vec![2], vec![1], vec![], true);
+    let exp_cs2 = conf_state(vec![1], vec![2]);
+
     let s = new_storage();
     let mut raw_node = new_raw_node(1, vec![1], 10, 1, s.clone(), &l);
     raw_node.campaign().expect("");
     let mut proposed = false;
-    let mut last_index;
-    let mut ccdata = vec![];
-    loop {
+    let ccdata = test_cc.write_to_bytes().unwrap();
+    // Propose the ConfChange, wait until it applies, save the resulting ConfState.
+    let mut cs = None;
+    while cs.is_none() {
         let rd = raw_node.ready();
         s.wl().append(rd.entries()).expect("");
-        // Once we are the leader, propose a command and a ConfChange.
-        if !proposed && rd.ss().is_some() && rd.ss().unwrap().leader_id == raw_node.raft.id {
-            raw_node.propose(vec![], b"somedata".to_vec()).expect("");
+        for e in rd.committed_entries.as_ref().unwrap() {
+            if e.get_entry_type() == EntryType::EntryConfChangeV2 {
+                let mut cc = ConfChangeV2::default();
+                cc.merge_from_bytes(e.get_data()).unwrap();
 
-            let cc = conf_change(ConfChangeType::AddNode, 2);
-            ccdata.reserve_exact(cc.compute_size() as usize);
-            cc.write_to_vec(&mut ccdata).unwrap();
-            raw_node.propose_conf_change(vec![], cc).expect("");
+                // Force it step down.
+                let mut msg = new_message(1, 1, MessageType::MsgHeartbeatResponse, 0);
+                msg.term = raw_node.raft.term + 1;
+                raw_node.step(msg).unwrap();
 
-            proposed = true;
+                cs = Some(raw_node.apply_conf_change(&cc).unwrap());
```

Comment:
Here, the leader step down first then enter joint, should we also test that a leader enter joint then step down?

Reply:
Enter joint then step down is the same as L198~L203. In normal flow, an empty entry is proposed immediately when entering joint state.